scripts/rollup/build.js

```diff
@@ -344,22 +385,28 @@ function getPlugins(
   const forks = Modules.getForks(bundleType, entry, moduleType, bundle);
   const isProduction = isProductionBundleType(bundleType);
   const isProfiling = isProfilingBundleType(bundleType);
-  const isUMDBundle =
-    bundleType === UMD_DEV ||
-    bundleType === UMD_PROD ||
-    bundleType === UMD_PROFILING;
-  const isFBWWWBundle =
-    bundleType === FB_WWW_DEV ||
-    bundleType === FB_WWW_PROD ||
-    bundleType === FB_WWW_PROFILING;
-  const isRNBundle =
-    bundleType === RN_OSS_DEV ||
-    bundleType === RN_OSS_PROD ||
-    bundleType === RN_OSS_PROFILING ||
-    bundleType === RN_FB_DEV ||
-    bundleType === RN_FB_PROD ||
-    bundleType === RN_FB_PROFILING;
-  const shouldStayReadable = isFBWWWBundle || isRNBundle || forcePrettyOutput;
+
+  const {isUMDBundle, shouldStayReadable} = getBundleTypeFlags(bundleType);
+
+  const needsMinifiedByClosure = isProduction && bundleType !== ESM_PROD;
+
+  // Any other packages that should specifically _not_ have sourcemaps
+  const sourcemapPackageExcludes = [
+    // Having `//#sourceMappingUrl` in this file breaks `ReactDevToolsHooksIntegration-test.js`,
```

Comment:
Why? Can we fix the test to be more resilient instead?

Reply:
Not sure. It took me close to an hour just to figure out that having a sourcemap URL was even the _reason_ why that test was failing, and I have no idea _why_ that was the case.  Seemed to be something about imports or source parsing, maybe.

Per the comment, that file really doesn't need to have sourcemaps anyway, so it's reasonable to exclude it.

Reply:
I appreciate the time spent on this, but we generally try to understand each problem exhaustively before attempting a fix. In this case, it's important to understand why the test fails. Especially since the test has something to do with DevTools integration (which relies on parsing stacks), and might be flagging a real problem.

Even if it's not a real problem, imagine someone adding another package with similar logic, and then being confused why this logic works doesn't fail tests when it's inside `react-debug-tools`, but does elsewhere. Then _they_ would have to spend an hour only to find that `react-debug-tools` is somehow exceptional.

We do need to get to the bottom of the real cause.

Reply:
Okay, I'll dig into this further and see if I can understand what's going on.

Reply:
`isStateEditable` is calculated [here](https://github.com/facebook/react/blob/8310854cebba4d1f404e65ff2064825ee76d78e2/packages/react-debug-tools/src/ReactDebugHooks.js#L595) so you might want to check if the `primitive` value is affected by the presence of sourcemaps. Maybe something gets borked in the stack parsing logic.

Reply:
Maybe functions get different names and [this logic](https://github.com/facebook/react/blob/8310854cebba4d1f404e65ff2064825ee76d78e2/packages/react-debug-tools/src/ReactDebugHooks.js#L470-L485) doesn't quite work anymore. Or something similar.

Reply:
Yeah, I just made two Replays so I can see what's going on:

- Passing: https://app.replay.io/recording/reactdevtoolshooksintegration-passing-no-sourcemappingurl--1f62a48f-6389-40b8-af7c-4ec5ac7c7352
- Failing: https://app.replay.io/recording/reactdevtoolshooksintegration-failing-with-sourcemappingurl--a8c8e399-4185-4ef9-9cd1-153457894213

And the _only_ difference code-wise to make it pass is that I hand-edited `react-debug-tools.production.min.js` to add an extra space in the `//#sourceMappingUrl=` line so that it no longer gets matched against.

Reply:
First difference I see is that the `stateHook` object in the test differs after calling `ReactDebugTools.inspectHooksOfFiber(fiber)`.

- Passing: `{id: 0, isStateEditable: true, name: "useState"}`
- Failing: `{id: null, isStateEditable: false, name: ".useState"}`

(adding comments and print statements to those Replays as I go)

Reply:
Yeah, there's definitely differences in both the `hookStack` and `primitiveStack` arrays.

For example, the `primitiveStack` object at index 1 is different. When passing, it's `{line: 16, functionName: "Object.useState"}`. When failing, it's `{line: 104, functionName: "Object.error [as useState]"}`.

So then when we get into `isReactWrapper()`, we're comparing different strings, and end up returning different `primitiveIndex` values.

```
// Passing:
Comparing hook names:  Proxy.useState useState
Comparing hook names:  Object.useState useState

// Failing:
Comparing hook names:  Proxy.Error useState
Comparing hook names:  Proxy.Error useState
```

I need to step away, but I think this justifies the reasoning for excluding `react-debug-tools` from having a sourcemap.  Clearly that changes the stack traces that the test is expecting, and that's breaking the test.

Reply:
@gaearon given that having a sourcemap for that `react-debug-tools.production.min.js` changes the hook names and the call stacks, does that justify excluding `react-debug-tools` from the sourcemaps list?

Reply:
I don't feel like I understand the conclusion.

What is `Proxy.Error`? Where is this defined? Why do we see this function name when the sourcemaps are applied? I understand we _do_ but shouldn't we answer the question of _why_?

>given that having a sourcemap for that react-debug-tools.production.min.js changes the hook names and the call stacks, does that justify excluding react-debug-tools from the sourcemaps list?

I don't think it does until we understand why this happens and what the actual root cause is. Is there a bug in `ErrorStackParser`? Is something throwing? We should fix it there then.

Reply:
Okay, I'll try to dig into this further. Are there any other concerns beyond this question that would be blocking this PR?  Would like to get a sense of how close this is to being acceptable.

Reply:
Did some further poking at this.  I _think_ this is due to the `react-debug-tools` logic combined with the actual test setup in the first place.

The stack trace string, prior to being parsed, is:

```
Error: 
    at Object.apply (/home/acemarke/projects/react/scripts/jest/setupTests.js:219:29)
    at Proxy.Error (/home/acemarke/projects/react/build/oss-stable/react-debug-tools/cjs/react-debug-tools.production.min.js:122:21)
    at Object.useReducer (/home/acemarke/projects/react/build/oss-stable/react/cjs/react.production.min.js:959:21)
    at MyComponent (/home/acemarke/projects/react/packages/react-debug-tools/src/__tests__/ReactDevToolsHooksIntegration-test.js:101:13)
    at renderFunction (/home/acemarke/projects/react/build/oss-stable/react-debug-tools/cjs/react-debug-tools.production.min.js:652:5)
    at Object.inspectHooks [as inspectHooksOfFiber] (/home/acemarke/projects/react/build/oss-stable/react-debug-tools/cjs/react-debug-tools.production.min.js:763:12)
    at Object.<anonymous> (/home/acemarke/projects/react/packages/react-debug-tools/src/__tests__/ReactDevToolsHooksIntegration-test.js:118:32)
```

Note that `Proxy.Error` is already in the stack trace string.

`react-debug-tools.production.min.js:122` is:

```ts
        useInsertionEffect: function(a) {
          C();
          x.push({
            primitive: "InsertionEffect",
            stackError: Error(),
            value: a
          })
        },
```

Note the line `stackError: Error()`, and `.stackError` is what's getting passed to the `error-stack-parser` library.

If you look at `scripts/jest/setupTests.js`, it has this logic that overwrites `global.Error` with a proxy:

```ts
    const ErrorProxy = new Proxy(OriginalError, {
      apply(target, thisArg, argumentsList) {
        const error = Reflect.apply(target, thisArg, argumentsList);
        error.message = decodeErrorMessage(error.message);
        return proxyErrorInstance(error);
      },
      construct(target, argumentsList, newTarget) {
        const error = Reflect.construct(target, argumentsList, newTarget);
        error.message = decodeErrorMessage(error.message);
        return proxyErrorInstance(error);
      },
      get(target, key, receiver) {
        if (key === 'captureStackTrace') {
          return captureStackTrace;
        }
        return Reflect.get(target, key, receiver);
      },
    });
    ErrorProxy.OriginalError = OriginalError;
    global.Error = ErrorProxy;
```

So, I think the inclusion of the sourcemap for `react-debug-tools.production.min.js` is causing Node to alter the stack trace related to that `global.Error = ErrorProxy`, which ends up breaking the parsing later.

Reply:
Sorry, I still don't quite understand. Where is the _mistake_?

Reply:
@gaearon : there is no "mistake" per se.

The test setup code works as intended.  The error stack parsing code works as intended.

But per my last comment, having a sourcemap for `react-debug-tools.production.min.js` causes Node to have different stack trace contents for the thrown error, which then causes the hook name parsing logic to fail to match what it expected to see.

We know this test worked fine when there was no sourcemap.  It's an internal package, and there's no _reason_ to have a sourcemap for this bundle.

So, we should skip it and keep the behavior the same as it was.

Reply:
> We know this test worked fine when there was no sourcemap. It's an internal package, and there's no reason to have a sourcemap for this bundle.

Can confirm that this package is a part of React DevTools' backend, but for historical reasons (which I don't know) it is present as a separate package in this monorepo.

Reply:
Yeah, I did just actually happen to talk to Brian today about the fact that gets used as part of the RDT backend to parse hook names.

That said, while it's technically published ( https://www.npmjs.com/package/react-debug-tools ) , it doesn't look like it's even been published at all since the first release 5 years ago, and it seems like the only references are inside of the RDT internals.  So, still no reason that I can see to justify having a sourcemap generated for that package, especially given the test situation.